In [26]:
import pandas as pd

from sklearn.metrics import accuracy_score
from sklearn.metrics import precision_score
from sklearn.metrics import recall_score
from sklearn.metrics import f1_score
from sklearn.metrics import roc_auc_score
from sklearn.metrics import precision_recall_curve

1) Загрузите файл classification.csv. В нем записаны истинные классы объектов выборки (колонка true) и ответы некоторого классификатора (колонка pred).

In [16]:
data = pd.read_csv(r'/Users/u17491440/PycharmProjects/enterInML/data/week_3/classification.csv')

data.head()

true  pred
0     1     0
1     1     1
2     1     1
3     0     0
4     1     1

2) Подсчитайте величины TP, FP, FN и TN согласно их определениям. Например, FP — это количество объектов, имеющих класс 0, но отнесенных алгоритмом к классу 1. Ответ в данном вопросе — четыре числа через пробел.

In [10]:
TP = data.query('true==1 & pred==1').shape[0]
FP = data.query('true==0 & pred==1').shape[0]
FN = data.query('true==1 & pred==0').shape[0]
TN = data.query('true==0 & pred==0').shape[0]
print(TP, FP, FN, TN, sep = ' ')

43 34 59 64


3) Посчитайте основные метрики качества классификатора:
- Accuracy (доля верно угаданных) — sklearn.metrics.accuracy_score
- Precision (точность) — sklearn.metrics.precision_score
- Recall (полнота) — sklearn.metrics.recall_score
- F-мера — sklearn.metrics.f1_score

В качестве ответа укажите эти четыре числа через пробел.

In [17]:
#Accuracy (доля верно угаданных) — sklearn.metrics.accuracy_score
print( accuracy_score(data['true'], data['pred']) )

#Precision (точность) — sklearn.metrics.precision_score
print( precision_score(data['true'], data['pred']) )

#Recall (полнота) — sklearn.metrics.recall_score
print( recall_score(data['true'], data['pred']) )

#F-мера — sklearn.metrics.f1_score
print( f1_score(data['true'], data['pred']) )

0.535
0.5584415584415584
0.4215686274509804
0.48044692737430167


4) Имеется четыре обученных классификатора. В файле scores.csv записаны истинные классы и значения степени принадлежности положительному классу для каждого классификатора на некоторой выборке:
- для логистической регрессии — вероятность положительного класса (колонка score_logreg),
- для SVM — отступ от разделяющей поверхности (колонка score_svm),
- для метрического алгоритма — взвешенная сумма классов соседей (колонка score_knn),
- для решающего дерева — доля положительных объектов в листе (колонка score_tree).

Загрузите этот файл.

In [23]:
data_scores = pd.read_csv(r'/Users/u17491440/PycharmProjects/enterInML/data/week_3/scores.csv')

data_scores.head()

true  score_logreg  score_svm  score_knn  score_tree
0     0      0.683832   0.145976   0.787063    0.500000
1     1      0.801966   0.239511   1.000000    0.833333
2     0      0.382315  -0.245701   0.000000    0.000000
3     1      0.506797  -0.137058   0.000000    0.105263
4     1      0.488781  -0.154148   0.000000    0.105263

5) Посчитайте площадь под ROC-кривой для каждого классификатора. Какой классификатор имеет наибольшее значение метрики AUC-ROC (укажите название столбца)? Воспользуйтесь функцией sklearn.metrics.roc_auc_score.

In [44]:
# ROC AUC score для каждого столбца
# Столбец первый
print( roc_auc_score(data_scores['true'], data_scores['score_logreg']), 'score_logreg')

# Столбец второй
print( roc_auc_score(data_scores['true'], data_scores['score_svm']), 'score_svm')

# Столбец третий
print( roc_auc_score(data_scores['true'], data_scores['score_knn']), 'score_knn')

# Столбец четвертый
print( roc_auc_score(data_scores['true'], data_scores['score_tree']), 'score_tree')

0.719187675070028 score_logreg
0.7086834733893557 score_svm
0.6351540616246498 score_knn
0.6919267707082833 score_tree


6) Какой классификатор достигает наибольшей точности (Precision) при полноте (Recall) не менее 70% ? 

In [43]:
# Наибольшая точность классификатора при полноте не меньше 70%
precision, recall, thresholds = precision_recall_curve(data_scores['true'].values, data_scores['score_logreg'].values)
precision_max = 0
classiff = ''
for j in ['score_logreg', 'score_svm', 'score_knn', 'score_tree']:
    precision, recall, thresholds = precision_recall_curve(data_scores['true'].values, data_scores[j].values)
    for i in range(len(precision)):
        if (recall[i]>0.7) and (precision[i]>precision_max):
            precision_max = precision[i]
            classiff = j
            print(classiff, precision_max, recall[i], sep=" ")
classiff

score_logreg 0.49746192893401014 1.0
score_logreg 0.5 0.9591836734693877
score_logreg 0.5026737967914439 0.9591836734693877
score_logreg 0.5027322404371585 0.9387755102040817
score_logreg 0.5054945054945055 0.9387755102040817
score_logreg 0.5082872928176796 0.9387755102040817
score_logreg 0.5111111111111111 0.9387755102040817
score_logreg 0.5139664804469274 0.9387755102040817
score_logreg 0.5168539325842697 0.9387755102040817
score_logreg 0.519774011299435 0.9387755102040817
score_logreg 0.5227272727272727 0.9387755102040817
score_logreg 0.5257142857142857 0.9387755102040817
score_logreg 0.5287356321839081 0.9387755102040817
score_logreg 0.5317919075144508 0.9387755102040817
score_logreg 0.5348837209302325 0.9387755102040817
score_logreg 0.5380116959064327 0.9387755102040817
score_logreg 0.5384615384615384 0.9285714285714286
score_logreg 0.5416666666666666 0.9285714285714286
score_logreg 0.5449101796407185 0.9285714285714286
score_logreg 0.5481927710843374 0.9285714285714286
score_logr

'score_tree'

1) Чему равны значения TP, FP, FN, TN (укажете четыре числа через пробел)?
43 34 59 64

2) Чему равны доля правильных ответов, точность, полнота, F-мера (укажите через пробел)?
0.54 0.56 0.42 0.48

3) Какой из классификаторов имеет наибольшее значение метрики AUC-ROC (укажите название столбца)?
score_logreg

4) Какой классификатор достигает наибольшей точности при полноте не менее 70% (укажите название столбца)?
score_tree